services/QuillLMS/lib/rule_feedback_history.rb

```diff
@@ -1,38 +1,70 @@
 class RuleFeedbackHistory
     def self.generate_report(conjunction:, activity_id:)
         sql_result = exec_query(conjunction: conjunction, activity_id: activity_id)
-        format_sql_results(sql_result)
+        postprocessed = postprocessing(sql_result)
+        format_sql_results(postprocessed)
     end
 
     def self.exec_query(conjunction:, activity_id:)
         sql = <<~SQL
           select rules.uid as rules_uid, 
+            MAX(feedback_histories.id),
             MAX(prompts.activity_id) as activity_id, 
             MAX(rules.rule_type) as rule_type, 
             MAX(rules.suborder) as rule_suborder,
             MAX(rules.name) as rule_name, 
             MAX(rules.description) as rule_description, 
-            COUNT(feedbacks.id) as feedback_histories_count
+            ARRAY_AGG(feedback_histories.id) as feedback_histories_id_array
+
           FROM comprehension_rules as rules 
           INNER JOIN comprehension_prompts_rules as prompts_rules ON rules.id = prompts_rules.rule_id 
           INNER JOIN comprehension_prompts as prompts ON prompts_rules.prompt_id = prompts.id 
-          LEFT JOIN feedback_histories as feedbacks ON feedbacks.rule_uid = rules.uid
+          LEFT JOIN feedback_histories ON feedback_histories.rule_uid = rules.uid
           WHERE prompts.conjunction = '#{conjunction}' AND activity_id = '#{activity_id}'
           GROUP BY rules_uid 
         SQL
         rules = Comprehension::Rule.find_by_sql(sql)
     end
 
     def self.postprocessing(rules_sql_result)
+        rules_sql_result.each do |r| 
+            
+            feedback_histories = FeedbackHistory
+                .where(id: r.feedback_histories_id_array).includes(:feedback_history_ratings)
+
+            total_scored = feedback_histories.reduce(0) do |memo, n|
+                memo + n.feedback_history_ratings.count
+            end
+
+            if total_scored == 0 
+                total_strong = 0
+                r.define_singleton_method(:pct_strong) { 0 }
+                r.define_singleton_method(:pct_scored) { 0 }
+            else 
+                total_strong = feedback_histories.reduce(0) do |memo, n|
+                    memo + n.feedback_history_ratings.filter(&:rating).count
+                end
+
+                r.define_singleton_method(:pct_strong) { total_strong/total_scored.to_f }
```

Comment:
I don't have alternative suggestions, but having a bunch of `define_singleton_method` calls in here makes me think there might be a cleaner approach?  But I can see the advantage of keeping the values in `rules_sql_result` as the same object type that was passed in since this is a post-processing script.  So we can go with this.

Reply:
Yeah, I was looking around for a way to merge an object with a struct, but didn't find a way.